design-documents/csp.md

```diff
@@ -0,0 +1,208 @@
+}
+```
+ 
+* Render given policy and apply to given response object.
+```php
+namespace Magento\Security\Model\CSP;
+
+interface PolicyRendererInterface
+{
+    public function render(PolicyInterface $policy, \Magento\Framework\App\Response\HttpInterface $response): void
+}
+```
+ 
+* A violation recorded by a browser
+```php
+namespace Magento\Security\Model\CSP\Data;
+
+interface ViolationInterface
+{
+    public function getDocumentUri(): string;
+    
+    public function getViolatedDirective(): string;
+    
+    public function getOriginalPolicy(): string;
+    
+    public function getReferrer(): ?string;
+    
+    public function getBlockedUri(): ?string;
+    
+    public function getSourceJsFileLine(): ?string;
+    
+    public function getReported(): \DateTime
+}
+```
+ 
+* A repository of violations logs collected when CSP reporting is enabled
+```php
+namespace Magento\Security\Model\CSP;
+
+interface ViolationsRepositoryInterface
+{
+    public function getList(SearchCriteriaInterface $search): ViolationSearchResultInterface;
+    
+    public function create(string $reportJson): void;
+}
+```
+ 
+## CSP configuration
+Security settings like CSP should be only manageable with physical files and
+CLI requiring highest level of access.
+ 
+CSP mode will be set with a config path not appearing in the admin configuration page so merchants
+would have to execute a CLI command in order to change it.
+
+## Nonce
+`script-src` and `style-src` directives allow to use _nonce_ to mark trusted _\<script\>_ and _\<style\>_ blocks which
+enables us to effectively disable `unsafe-inline` except for trusted blocks which may make migration to CSP friendly
+front-end easier for Magento core and merchants alike. Nonce can be used for script/style blocks containing JS/CSS but
+also for the ones having `src` attribute defined. Using this mechanism merchants can whitelist external scripts with
+nonce instead of having to add every external domain to the whitelist.
+ 
+There is no way to whitelist an event handler defined via an attribute though (like _onclick_) or styles defined
+with _style_ attribute that's why nonce will be
+disabled by default and all `unsafe-inline` scripts/styles will be allowed. Merchants who'll be successful in getting
+read of all event handlers/styles defined using HTML attributes will be able to enable _nonce_ via configuration
+and used it for whitelisting. With _nonce_ enable magento will also set `strict-dynamic` value for _script-src_ policy
+to allow scripts whitelisted with _nonce_ to load other scripts. This will be crucial to allow analytics/advertisement
+scripts to work since they often add a bunch of other scripts from a variety of domains when loaded.
+ 
+It is possible to combine nonce and domain-list based whitelisting for CSP.
+ 
+Magento will have to generate unique nonce (using _Magento\Framework\Math\Random_) for each response and developers will
```

Comment:
Is `Magento/Framework/Math/Random` cryptographically secure?

Reply:
Yes, at the core it uses `random_int` which is.